### PR TITLE
[warning] Add a warning with missing callback with resume_from_checkpoint

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Add a warning with missing callback with resume_from_checkpoint

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/callback_hook.py [hunk 1]
```diff
@@ -293,10 +293,22 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
...
+        callback_states = checkpoint.get('callbacks')
+
+        current_callbacks_type = {type(cb) for cb in self.callbacks}
+        saved_callbacks_type = set(callback_states)
+        difference = saved_callbacks_type.difference(current_callbacks_type)
+        if difference:
+            rank_zero_warn(
+                "Be aware that when using ``resume_from_checkpoint``, "
+                "callbacks used to create the checkpoint need to be provided. "
+                f"Please, add the following callbacks: {list(difference)}. ", UserWarning
```
Comment (awaelchli) [inline pytorch_lightning/trainer/callback_hook.py:309]:
```suggestion
                "Be aware that when using ``resume_from_checkpoint``,"
                " callbacks used to create the checkpoint need to be provided."
                f" Please add the following callbacks: {list(difference)}. ", UserWarning
```

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -2047,3 +2047,24 @@ def test_fit_test_synchronization(tmpdir):
...
+
+
+class CustomCallbackOnLoadCheckpoint(Callback):
+
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint) -> dict:
+        return {"a": None}
+
+
+def test_on_load_checkpoint_missing_callbacks(tmpdir):
+
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:2059]:
```suggestion
def test_on_load_checkpoint_missing_callbacks(tmpdir):
    """ Test a warning appears when callbacks in the checkpoint don't match callbacks provided when resuming. """
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/callback_hook.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/callback_hook.py:309]: is this warning always needed? For example, i may have saved a checkpoint dict which contained callback states from the model checkpoint. But when I resume from checkpoint, I might not set a new model checkpoint callback on the trainer 

is there an issue/question which prompted the PR?

Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:309]: This PR checks the `AU~B` between `save_callback_state` and `current_callback_state`.
Basically, if you trained with a `CustomCallback` and you don't provide it back when `resuming`, 
we won't be able to reload the state as this callback isn't provided. 
Therefore we should provide a warning that `resume` might not have the expected behaviour.

Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:309]: `resume_from_checkpoint` is bringing lot of confusion as we expect exactly the same arguments to the Trainer as provided when training before.

File: pytorch_lightning/trainer/callback_hook.py [hunk 3]
```diff
@@ -293,10 +293,22 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
...
         """Called when loading a model checkpoint."""
-        callback_states = checkpoint.get('callbacks')
+
         # Todo: the `callback_states` are dropped with TPUSpawn as they
         # can't be saved using `xm.save`
         # https://github.com/pytorch/xla/issues/2773
+        callback_states = checkpoint.get('callbacks')
+
+        current_callbacks_type = {type(cb) for cb in self.callbacks}
+        saved_callbacks_type = set(callback_states)
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/callback_hook.py:303]:
```suggestion
        saved_callbacks_type = set(callback_states.keys())
```

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley: maybe just do `set(callback_states.keys())` for readability

**Review by SeanNaren — APPROVED** [review]

**Pull request merged**